src/Compilers/CSharp/Portable/CodeGen/CodeGenerator_RefSafety.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeGen;
+
+internal partial class CodeGenerator
+{
+    private static bool MightEscapeTemporaryRefs(BoundCall node, bool used, AddressKind? receiverAddressKind)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: node.Method.RefKind,
+            receiverType: !node.Method.RequiresInstanceReceiver ? null : node.ReceiverOpt?.Type,
+            receiverScope: node.Method.TryGetThisParameter(out var thisParameter) ? thisParameter?.EffectiveScope : null,
+            receiverAddressKind: receiverAddressKind,
+            isReceiverReadOnly: node.Method.IsEffectivelyReadOnly,
+            parameters: node.Method.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(BoundObjectCreationExpression node, bool used)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: RefKind.None,
+            receiverType: null,
+            receiverScope: null,
+            receiverAddressKind: null,
+            isReceiverReadOnly: false,
+            parameters: node.Constructor.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(BoundFunctionPointerInvocation node, bool used)
+    {
+        FunctionPointerMethodSymbol method = node.FunctionPointer.Signature;
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: method.RefKind,
+            receiverType: null,
+            receiverScope: null,
+            receiverAddressKind: null,
+            isReceiverReadOnly: false,
+            parameters: method.Parameters);
+    }
+
+    private static bool MightEscapeTemporaryRefs(
+        bool used,
+        TypeSymbol returnType,
+        RefKind returnRefKind,
+        TypeSymbol? receiverType,
+        ScopedKind? receiverScope,
+        AddressKind? receiverAddressKind,
+        bool isReceiverReadOnly,
+        ImmutableArray<ParameterSymbol> parameters)
+    {
+        Debug.Assert(receiverAddressKind is null || receiverType is not null);
+
+        // We check the signature of the method, counting potential `ref` sources and destinations
+        // to determine whether a `ref` can be captured by the method.
+        // The emit layer then uses this information to avoid reusing temporaries that are passed by ref to such methods.
+
+        // number of outputs that can capture `ref`s
+        int refTargets = 0;
+        // number of inputs that can contain `ref`s
+        int refSources = 0;
```

Comment:
Just curious: Why count ref inputs and outputs rather than simply checking if there are any of each?

Reply:
The logic used to be more complicated before I realized silly cyclic assignment needs to be handled and then simplified this, so yeah, you're right now booleans should be enough, I will change that, thanks.